INT-4064: Simplify `IdempotentReInt` Java Config

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4064

We configure `IdempotentReceiver` via `<int:idempotent-receiver>` component or `@IdempotentReceiver` annotation.
In case of regular Java config, e.g. direct `ConsumerEndpointFactoryBean` usage or Java DSL,
it isn't possible to configure `idempotentReceiverInterceptor` enough easy
- Introduce `if...else` logic into `ConsumerEndpointFactoryBean` to proxy `MessageHandler`,
  if `adviceChain` contains an newly-introduced `HandleMessageAdvice`.
  And do that independently if `MessageHandler` is `AbstractReplyProducingMessageHandler`
- Make `idempotentReceiverInterceptor extends HandleMessageAdvice`
- Skip `HandleMessageAdvice` in the `AbstractReplyProducingMessageHandler`
- Add advice applying logic into the `AbstractMethodAnnotationPostProcessor` as well

artembilan committed:
INT-4064: Simplify `IdempotentReInt` Java Config

JIRA: https://jira.spring.io/browse/INT-4064

We configure `IdempotentReceiver` via `<int:idempotent-receiver>` component or `@IdempotentReceiver` annotation.
In case of regular Java config, e.g. direct `ConsumerEndpointFactoryBean` usage or Java DSL,
it isn't possible to configure `idempotentReceiverInterceptor` enough easy

* Introduce `if...else` logic into `ConsumerEndpointFactoryBean` to proxy `MessageHandler`,
if `adviceChain` contains an newly-introduced `HandleMessageAdvice`.
And do that independently if `MessageHandler` is `AbstractReplyProducingMessageHandler`
* Make `idempotentReceiverInterceptor extends HandleMessageAdvice`
* Skip `HandleMessageAdvice` in the `AbstractReplyProducingMessageHandler`
* Add advice applying logic into the `AbstractMethodAnnotationPostProcessor` as well

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/config/ConsumerEndpointFactoryBean.java:
```diff
@@ -185,30 +186,30 @@ public void afterPropertiesSet() throws Exception {
+			boolean replyMessageHandler = AbstractReplyProducingMessageHandler.class.isAssignableFrom(targetClass);
+
+			for (Advice advice : this.adviceChain) {
+				if (!replyMessageHandler || advice instanceof HandleMessageAdvice) {
```
The advice itself might be a proxy - need an interface.

garyrussell commented:
I am uneasy about this change...

- it is not at all clear why the idempotent receiver advice,  when used on an `ARPMH`, needs to advise the entire flow instead of just the `handleRequestMessage` method.
- this changes the advice order so the `IdempotentReceiverInterceptor` will always be applied first, regardless of its position in the advice chain (with an `ARPMH`) - this means the behavior will not be as expected, and is a breaking change so not permissible in a point release.

artembilan commented:
1. See the `IdempotentReceiverInterceptor` logic. There is a case when we return `null` instead of `invocation.proceed()`. Therefore in case of `handleRequestMessage` advising we end up with the:
```
if (result != null) {
	sendOutputs(result, message);
}
else if (this.requiresReply && !isAsync()) {
	throw new ReplyRequiredException(message, "No reply produced by handler '" +
	getComponentName() + "', and its 'requiresReply' property is set to true.");
}
```
Which doesn't fit to premise of:
```
* If there is needed just silently 'drop' rejected messages configure the
* {@link #discardChannel} to the {@code nullChannel}.
```
That's why it felt like a `filter` from the beginning ( #1286 ), because we shouldn't allow endpoint to pass message to downstream if that is a goal, but we should drop it without any effect, because it is noisy duplicate. 
According to your [conclusion](https://github.com/spring-projects/spring-integration/pull/1286#issuecomment-59394254) there, we decided to end up with an `Advice`. But since we are not interested in the `ReplyRequiredException`, we advise the whole `handleMessage()`, instead of just internal `handleRequestMessage`: 

2. Not sure where you see the order change, because `IdempotentReceiverInterceptor` isn't designed for anything else unless:
```
boolean isMessageHandler = invocationThis != null && invocationThis instanceof MessageHandler;
boolean isMessageMethod = method.getName().equals("handleMessage")
		&& (arguments.length == 1 && arguments[0] instanceof Message);
```
So, it was ignored previously in the `<request-handler-advice-chain>` anyway, but its own internal logic. Since now we pushed our hands into `ARPMH` guts, we are free to ignore the interceptor on the init phase already.
If it isn't `ARPMH`, anything has been changed and we apply all the advises for the `handleMessage`.
OTOH `IRI` has been applied as first one for the `handleMessage` previously anyway.
See `@IdempotentReceiver` and its `IdempotentReceiverAutoProxyCreator`.

artembilan committed:
Introduce `HandleMessageAdvice` marker interceptor to cover the case when an `Advice` can be advices as well.
Remove unused `setAdviceChainIfPresent()` method in the `AbstractMethodAnnotationPostProcessor`

garyrussell commented:
Ugh - unintended consequences 😦 

I see, but we need to document the restriction.

Someone might (perhaps reasonably) configure a transaction interceptor in the advice chain and expect the IRI and handler to run in the tx.

artembilan commented:
> Someone might (perhaps reasonably) configure a transaction interceptor in the advice chain and expect the IRI and handler to run in the tx.

??? Even if he would do that, the `IRI` doesn't work anyway, because it isn't going to perform its logic over `handleRequestMessage`.

I won't mind to add something to docs, but need to understand what you mean. For example there is no strict message in the [Provided Advice Classes](http://docs.spring.io/spring-integration/docs/4.3.1.BUILD-SNAPSHOT/reference/html/messaging-endpoints-chapter.html#advice-classes) that `AbstractRequestHandlerAdvice` is only for `handleRequestMessage` and `handleMessage`. We can add a note, of course, too. Just let me know!

Regarding order for `@Transactional`, `@IdempotentReceiver`, `@Async`, `@Scheduled`, `@Validated`, `@Publisher` etc., I'm not sure what to say you. I don't see any clues in the source code in which order they are applied to the target object. We can ask Juergen and Stephane what's going on, but I feel that answer would be to use manual proxy creation via direct advice applying, similar to our `adviceChain` option. Or... you all those annotation on different methods for your call stack and don't apply all of them on one top-level service method.

garyrussell commented:
Doh - I forgot that it's a top-level element with a pattern; I was thinking it was configured in XML like this:

    <int:request-handler-advice-chain>
         <tx:advice ... />
         <int:idempotent-receiver ... />
    </request-handler-advice-chain>

We probably need to explicitly prevent this configuration - even if the namespace doesn't support it as a nested element, someone could try to wire it up this way as a bean which, of course, wouldn't work as expected (before or after this change).

artembilan commented:
Well, actually this change exactly provides the ability `to wire it up this way as a bean` and `ConsumerEndpointFactoryBean` will properly distribute Advices.

OK. Let me add a note into Docs that `HandleMessageAdvice` s like `IdempotentReceiverInterceptor` are not applied to the `handleRequestMessage`.

Is that fine? 

Thank you for your nit-picking anyway :smile:

garyrussell commented:
>Well, actually this change exactly provides the ability to wire it up this way as a bean and ConsumerEndpointFactoryBean will properly distribute Advices.

Yes, but it won't be intuitive that the interceptor is not part of the transaction.

artembilan committed:
Document `HandleMessageAdvice`

artembilan committed:
Increase wait latch timeouts in the `LockRegistryLeaderInitiatorTests`

garyrussell commented:
Doc polishing here https://github.com/garyrussell/spring-integration/commit/250b8e6268e2e5eac5faf846919d1797a832fd53 for your approval before merge.

artembilan commented:
OK. Good with me. 
Thank you!

garyrussell commented:
Merged as d9f9f9e3ca77473b92f7017ddc33c791fb702f1c
